src/flow.ts

```diff
@@ -1351,6 +1351,13 @@ export class Flow {
           case UnaryOp.ClzI32:
           case UnaryOp.CtzI32:
           case UnaryOp.PopcntI32: return type.size < 7;
+
+          // sign extensions overflow if result can have high garbage bits in the target type
+          case UnaryOp.Extend8I32: return type.size < (type.isUnsignedIntegerValue ? 32 : 8);
+          case UnaryOp.Extend8I64: return type.size < (type.isUnsignedIntegerValue ? 64 : 8);
+          case UnaryOp.Extend16I32: return type.size < (type.isUnsignedIntegerValue ? 32 : 16);
+          case UnaryOp.Extend16I64: return type.size < (type.isUnsignedIntegerValue ? 64 : 16);
+          case UnaryOp.Extend32I64: return type.size < (type.isUnsignedIntegerValue ? 64 : 32);
```

Comment:
Perhaps a comment here for review, by example of `i32.extend16_s`:

If the target type is unsigned, an `i32.extend16_s` can yield high garbage bits if the target type is smaller than `u32`, which naturally does not care about garbage bits.

If the target type is signed, an `i32.extend16_s` can yield high garbage bits if the target type is smaller than `i16`. Otherwise it would set all high bits to either 0 or 1, which is valid.